File: 03-sliding-window/04-minimum-window-substring/Solution.py
from collections import Counter

class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if not s or not t:
            return ""
        need = Counter(t)
        missing = len(t)
        l = 0
        best_l, best_r = 0, 0
        for r, c in enumerate(s, 1):
            if need[c] > 0:
                missing -= 1
            need[c] -= 1
            if missing == 0:
                while l < r and need[s[l]] < 0:
                    need[s[l]] += 1
                    l += 1
                if best_r == 0 or r - l < best_r - best_l:
                    best_l, best_r = l, r
                need[s[l]] += 1
                missing += 1
                l += 1
        return s[best_l:best_r]

File: 03-sliding-window/04-minimum-window-substring/Solution.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        if (s.empty() || t.empty()) return "";
        unordered_map<char,int> need;
        for (char c : t) need[c]++;
        int missing = t.size();
        int l = 0, bestL = 0, bestR = 0;
        for (int r = 1; r <= (int)s.size(); r++) {
            char c = s[r - 1];
            if (need[c] > 0) missing--;
            need[c]--;
            if (missing == 0) {
                while (l < r && need[s[l]] < 0) {
                    need[s[l]]++;
                    l++;
                }
                if (bestR == 0 || r - l < bestR - bestL) {
                    bestL = l; bestR = r;
                }
                need[s[l]]++;
                missing++;
                l++;
            }
        }
        return s.substr(bestL, bestR - bestL);
    }
};
